tx_test.go

```diff
@@ -10,14 +11,132 @@ import (
 	"testing"
 	"time"
 
-	"github.com/stretchr/testify/assert"
-	"github.com/stretchr/testify/require"
-
 	bolt "go.etcd.io/bbolt"
 	berrors "go.etcd.io/bbolt/errors"
 	"go.etcd.io/bbolt/internal/btesting"
+	"go.etcd.io/bbolt/internal/common"
+	"go.etcd.io/bbolt/internal/guts_cli"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
+func TestTx_Check_Page(t *testing.T) {
+	bucketKey := "testBucket"
+	db := btesting.MustCreateDBWithOption(t, &bolt.Options{PageSize: 4096})
+	defer db.MustDeleteFile()
+
+	// arrange
+	if err := db.Update(func(tx *bolt.Tx) error {
+		b, err := tx.CreateBucketIfNotExists([]byte(bucketKey))
+		if err != nil {
+			t.Fatal(err)
+		}
+		// insert random data
+		generateSampleDataInBucket(t, b, 3)
+		return nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+	db.MustCheck()
+
+	// act : corrupt a leaf page
+	victimPageId, _ := corruptLeafPage(t, db)
+
+	// assert: expected an error after check
+	if err := db.View(func(tx *bolt.Tx) error {
+		for err := range tx.Check() {
+			require.ErrorContains(t, err, fmt.Sprintf("leaf page(%d)", victimPageId))
+		}
+		return nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+	db.MustClose()
+}
+
+// corruptLeafPage write an invalid leafPageElement into the victim page.
+func corruptLeafPage(t testing.TB, db *btesting.DB) (victimPageId common.Pgid, validPageIds []common.Pgid) {
+	t.Helper()
+
+	victimPageId, validPageIds = findVictimPageId(t, db)
+
+	victimPage, victimBuf, err := guts_cli.ReadPage(db.Path(), uint64(victimPageId))
+	require.NoError(t, err)
+	require.NotNil(t, victimPage)
+	require.NotNil(t, victimBuf)
+	require.True(t, victimPage.IsLeafPage())
+	require.True(t, victimPage.Count() > 0)
+
+	// Insert LeafPageElement which violates the BTree range.
+	invalidLPE := common.NewLeafPageElement(0, 0, 0, 0)
+	var buf bytes.Buffer
+	binary.Write(&buf, binary.BigEndian, invalidLPE)
+	copy(victimBuf[16:], buf.Bytes())
+	// Write the corrupt page to db file.
+	err = guts_cli.WritePage(db.Path(), victimBuf)
+	require.NoError(t, err)
+
+	return victimPageId, validPageIds
+}
+
+// findVictimPageId finds all the leaf pages of a bucket and picks a leaf page to be corrupted.
+func findVictimPageId(t testing.TB, db *btesting.DB) (victimPageId common.Pgid, validPageIds []common.Pgid) {
```

Comment:
as discussed in https://github.com/etcd-io/bbolt/pull/659#issuecomment-1881033568 

- after generating the sample data, the DB's RootPage is no longer `pgid(4)`.
- I read the db's rootPage
- Then I find the bucket's rootPage (i.e. a branchPage)
- Then I traverse the `branchPageElements` finding the all the leafPageIds
- Then I corrupt the first leafPage

Tx.Check does not return the valid pages, only the corrupted pages